### PR TITLE
Porting of SlackAdapter class to C#

### PR DESCRIPTION
**Pull request by matiasroldan6**
## Description
SlackAdapter.cs and its class were ported from the original [slack_adapter.ts](https://github.com/southworkscom/botkit/blob/master/packages/botbuilder-adapter-slack/src/slack_adapter.ts) in Botkit js. The work is in progress and some of the code was unable to being implemented just yet.
This is a work in progress.

### CODE REVIEW

File: C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs [hunk 1]
```diff
@@ -100,7 +108,20 @@ public async Task<SlackAPI> GetAPI(Activity activity)
...
         {
-            return "";
+            if (!string.IsNullOrEmpty(Identity))
+            {
+                return Identity;
+            }
+            else if ((activity.Conversation as dynamic).team != null)
+            {
+                var userID = await options.GetBotUserByTeam((activity.Conversation as dynamic).team);
+                return string.IsNullOrEmpty(userID) ? userID : throw new Exception("Missing credentials for team.");
```
Comment (ceciliaavila) [inline C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs]: This should be 
```
return string.IsNullOrEmpty(userID) ? throw new Exception("Missing credentials for team.") : userID;
```

File: C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs [hunk 2]
```diff
@@ -119,7 +142,16 @@ public string GetInstallLink()
         /// <param name="code">The value found in `req.query.code` as part of Slack's response to the oauth flow.</param>
         public async Task<object> ValidateOauthCode(string code)
         {
-            return new object();
+            SlackAPI slack = new SlackAPI();
+            /*var*/ dynamic results = string.Empty; //await slack.oauth.access(); // TODO: Implement 'slack.oauth.access' in 'SlackApi'
+            if (results.ok)
```
Comment (VictorGrycuk) [inline C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs]: These lines could be simplified as
``` c#
return results.ok ? results : throw new Exception(results.error);
```
**If** `results.ok` is a `bool` type

File: C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs [hunk 3]
```diff
@@ -178,24 +311,249 @@ public async void ContinueConversation(ConversationReference reference, Action<T
...
+                        ChannelId = "slack",
+                        Conversation = new ConversationAccount()
+                        {
+                            Id = slackEvent.channel.id
+                            // thread_ts = slackEvent.thread_ts,
+                            // team = slackEvent.team.id
+                        },
+                        From = new ChannelAccount()
+                        {
+                            Id = slackEvent.bot_id ? slackEvent.bot_id : slackEvent.user.id
```
Comment (VictorGrycuk) [inline C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs]: This ternary should check if `slackEvent.bot_id` is not null, as `bot_id` is not a bool type

File: C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs [hunk 4]
```diff
@@ -109,93 +146,433 @@ public async Task<string> GetBotUserByTeam(Activity activity)
...
+                    SlackAPI slack = await GetAPIAsync(activity);
+                    //results = await slack.chat.update(message);
+                    if (/*!results.ok*/ true)
+                    {
+                        Console.WriteLine("Error updating activity on Slack:", results);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Console.WriteLine("Error updating activity on Slack:", ex.Message);
```
Comment (GasparAcevedoZainSouthworks) [inline C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs]: Replace the _Console.WriteLine_ with `throw new exception`

File: C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs [hunk 5]
```diff
@@ -109,93 +146,433 @@ public async Task<string> GetBotUserByTeam(Activity activity)
...
+            {
+                try
+                {
+                    SlackAPI slack = await GetAPIAsync(turnContext.Activity);
+                    // results = await slack.chat.delete({ ts: reference.activityId, channel: reference.conversation.id });
+                }
+                catch (Exception ex)
+                {
+                    Console.WriteLine("Error deleting activity", ex.Message);
+                    throw ex;
```
Comment (GasparAcevedoZainSouthworks) [inline C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs]: Same here, regarding the _console.writeline_

File: C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs [hunk 6]
```diff
@@ -109,93 +146,433 @@ public async Task<string> GetBotUserByTeam(Activity activity)
...
+                response.Content = new ObjectContent((slackEvent.challenge as object).GetType(), slackEvent.challenge, formatters[0]);
+                return;
+            }
 
-        public override Task<ResourceResponse> UpdateActivityAsync(ITurnContext turnContext, Activity activity, CancellationToken cancellationToken)
-        {
-            throw new NotImplementedException();
-        }
+            if (!VerifySignatureAsync(request, response))
+            {
```
Comment (GasparAcevedoZainSouthworks) [inline C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs]: Is this needed, as there isn't code lines

File: C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs [hunk 7]
```diff
@@ -109,93 +146,433 @@ public async Task<string> GetBotUserByTeam(Activity activity)
...
+                    }
+                    else
+                    {
+                        response.Content = new ObjectContent(typeof(string), string.Empty, formatters[0]);
+                    }
+                }
+            }
+            else
+            {
+                Console.WriteLine("Unknown Slack event type: ", slackEvent);
```
Comment (GasparAcevedoZainSouthworks) [inline C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs]: Throw exception

File: C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs (hunk 1 above)
Comment (ceciliaavila) [reply C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs]: This is still the same.

Comment (matiasroldan6) [reply C#/Botkit/Microsoft.BotKit.Adapters.Slack/SlackAdapter.cs]: I'm seeing this in the code
`return !string.IsNullOrEmpty(userID) ? userID : throw new Exception("Missing credentials for team.");`

**Pull request merged**